src/content/language/macros.md

```diff
@@ -0,0 +1,300 @@
+common issue of tedious JSON serialization in Dart.
+
+- [**The macros feature in general**](#the-macros-language-feature):
+Why we're adding macros to Dart, motivating use cases,
+benefits over existing code gen solutions,
+and a cursory overview of how writing macros will work in the future once
+the feature is complete.
+
+[spec]: {{site.repo.dart.lang}}/blob/main/working/macros/feature-specification.md
+[motivation]: {{site.repo.dart.lang}}/language/blob/main/working/macros/motivation.md
+
+## The `JsonCodable` macro
+
+:::important
+The `JsonCodable` macro is not stable and currently behind an [experiment flag][].
+It only works with the [dev channel][channel] releases of Dart 3.5.0 and later.
+
+Functionality is subject to change.
+:::
+
+The [`JsonCodable`][] macro encodes and decodes
+user-defined Dart classes to JSON maps of type `Map<String, Object?>`.
+It generates two members, a `toJson` serialization method,
+and a `fromJson` deserialization constructor.
+
+[experiment flag]: /tools/experiment-flags
+[`JsonCodable`]: {{site.pub-pkg}}/json/versions/0.20.0
+[channel]: https://dart.dev/get-dart#release-channels
+
+### Set up the experiment
+
+1. Switch to the [dev channel][channel] and upgrade to at least 3.5.0.
+
+2. [Add the package][] to your pubspec and retrieve
+  its dependencies.
+3. [Add the experiment][] to the `analysis_options.yaml`
+  file at the root of your project:
+
+   ```yaml
+   analyzer:
+    enable-experiment:
+      - macros
+   ```
+4. Import it in the file you plan to use it:
+
+   ```dart
+   import 'package:json/json.dart';
+   ```
+
+5. Run your project with the experiment flag:
+
+   ```console
+   dart --enable-experiment=macros run bin/my_app.dart
+   ```
+
+[Add the package]: /guides/packages
+[Add the experiment]: /tools/experiment-flags#using-experiment-flags-with-the-dart-analyzer-command-line-and-ide
+
+### Use the macro
+
+To use the `JsonCodable` macro, attach the annotation to the class you want to serialize:
+
+```dart
+import 'package:json/json.dart';
+
+@JsonCodable() // Macro annotation.
+class User {
+ final int? age;
+ final String name;
+ final String username;
+}
+```
+
+The macro introspects the `User` class and derives the implementations of
+`fromJson` and `toJson` using the `User` class's fields.
+
+So, without needing to define them yourself, `toJson` and `fromJson` are now
+available to use on objects of the annotated class:
+
+```dart
+void main() {
+ // Given some arbitrary JSON:
+ var userJson = {
+   'age': 5,
+   'name': 'Roger',
+   'username': 'roger1337'
+ };
+
+ // Use the generated members:
+ var user = User.fromJson(userJson);
+ print(user);
+ print(user.toJson());
+}
+```
+
+### View the generated code
+
+Sometimes it can be useful to view the generated code to better understand
+how a macros works, or to inspect the details of what it offers.
+
+Click on the "**Go to Augmentation**" link that appears under the annotation
+in your IDE (supported in VSCode and IntelliJ)
```

Comment:
I don't think we have it completed (or at least not tested) in IntelliJ. I think we should state just VSCode for now, and then offline hear if IntelliJ should be added.